File: papr/layouts/month.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Month Layout - Single month calendar on a portrait page.
Top 1/3: Year (small) and month abbreviation (large).
Bottom 2/3: 7-column grid (Mon-Sun) with weekday headers.
"""

import calendar
import datetime
import logging

import cairo
from gi.repository import Pango

from papr.util import drawing, metrics, styles


def drawCalendar(env):
    logging.debug(
        "Adding additional information to environment specific to Month layout"
    )

    # Portrait orientation - use dimensions as-is
    page_width = env.width
    page_height = env.height

    # Available area after margins
    available_width = page_width - (2 * env.safety)
    available_height = page_height - (2 * env.safety)

    # Split page: 1/3 header, 2/3 grid
    env.header_height = available_height / 3
    env.grid_height = available_height * 2 / 3

    # Grid setup: 7 columns (Mon-Sun)
    env.columns = 7

    # Calculate rows needed (max 6 weeks + 1 header row)
    env.header_row_height = env.grid_height * 0.08  # weekday header row
    env.grid_content_height = env.grid_height - env.header_row_height
    env.rows = 6  # max weeks in a month view
    env.cell_width = available_width / env.columns
    env.cell_height = env.grid_content_height / env.rows

    # Offsets
    env.offset_x = env.safety
    env.header_offset_y = env.safety
    env.weekday_header_y = env.safety + env.header_height
    env.grid_offset_y = env.weekday_header_y + env.header_row_height

    # Style settings from central config
    env.line_width = styles.BORDER_WIDTH * metrics.MM
    env.border_color = styles.BORDER_COLOR
    env.weekend_bg_color = styles.WEEKEND_BG_COLOR
    env.text_color_primary = styles.TEXT_COLOR_PRIMARY
    env.text_color_secondary = styles.TEXT_COLOR_SECONDARY
    env.text_color_muted = styles.TEXT_COLOR_MUTED

    # Font sizes
    env.day_text_size = 8
    env.weekday_header_font_size = env.header_row_height * 0.5

    logging.debug("Page: %s x %s mm", page_width / metrics.MM, page_height / metrics.MM)
    logging.debug("Header height: %s mm", env.header_height / metrics.MM)
    logging.debug("Grid: %s cols x %s rows", env.columns, env.rows)
    logging.debug(
        "Cell size: %s x %s mm",
        env.cell_width / metrics.MM,
        env.cell_height / metrics.MM,
    )

    # Create PDF surface (portrait)
    surface = drawing.create_surface(env.out, page_width, page_height)
    cr = cairo.Context(surface)

    # Draw components
    date = datetime.date(env.year, env.month, 1)
    drawHeader(cr, env, date)
    drawWeekdayHeaders(cr, env)
    drawDaysGrid(cr, env, date)

    logging.info("Finished drawing Calendar!")


def drawHeader(cr, env, date):
    """Draw year (small) and month abbreviation (large) in top 1/3."""
    with drawing.restoring_transform(cr):
        year_str = str(date.year)
        month_str = date.strftime("%b").upper()

        # Calculate font sizes - month is 5x year
        # Start with month taking ~60% of header height
        month_font_size = int(env.header_height * 0.5)
        year_font_size = int(month_font_size / 5)

        year_font = Pango.FontDescription("%s bold %s" % (env.font, year_font_size))
        month_font = Pango.FontDescription("%s bold %s" % (env.font, month_font_size))

        # Measure year text
        year_layout = drawing.create_layout_with_kerning(cr)
        year_layout.set_font_description(year_font)
        year_layout.set_text(year_str, -1)
        year_width, year_height = year_layout.get_pixel_size()

        # Measure month text
        month_layout = drawing.create_layout_with_kerning(cr)
        month_layout.set_font_description(month_font)
        month_layout.set_text(month_str, -1)
        month_width, month_height = month_layout.get_pixel_size()

        # Center horizontally
        available_width = env.width - (2 * env.safety)
        center_x = env.offset_x + available_width / 2

        # Center month in header area (vertically and horizontally)
        month_y = env.header_offset_y + (env.header_height - month_height) / 2
        month_x = center_x - month_width / 2

        # Place year above month with fixed padding
        year_month_gap = year_height * 0.3
        year_y = month_y - year_month_gap - year_height
        year_x = center_x - year_width / 2

        # Ensure year doesn't go above the top margin
        if year_y < env.header_offset_y:
            year_y = env.header_offset_y

        # Draw year (above month)
        cr.move_to(year_x, year_y)
        with drawing.layout(cr) as layout:
            layout.set_font_description(year_font)
            layout.set_text(year_str, -1)
            cr.set_source_rgb(*env.text_color_muted)

        # Draw month (centered in header)
        cr.move_to(month_x, month_y)
        with drawing.layout(cr) as layout:
            layout.set_font_description(month_font)
            layout.set_text(month_str, -1)
            cr.set_source_rgb(*env.text_color_primary)


def drawWeekdayHeaders(cr, env):
    """Draw weekday abbreviations as column headers (MON TUE WED THU FRI SAT SUN)."""
    weekdays = ["MON", "TUE", "WED", "THU", "FRI", "SAT", "SUN"]

    font = Pango.FontDescription(
        "%s bold %s" % (env.font, int(env.weekday_header_font_size))
    )

    for col, day_name in enumerate(weekdays):
        x = env.offset_x + (col * env.cell_width)
        y = env.weekday_header_y

        with drawing.restoring_transform(cr):
            cr.translate(x, y)

            # Measure text for centering
            temp_layout = drawing.create_layout_with_kerning(cr)
            temp_layout.set_font_description(font)
            temp_layout.set_text(day_name, -1)
            text_width, text_height = temp_layout.get_pixel_size()

            # Center in cell
            text_x = (env.cell_width - text_width) / 2
            text_y = (env.header_row_height - text_height) / 2

            cr.move_to(text_x, text_y)
            with drawing.layout(cr) as layout:
                layout.set_font_description(font)
                layout.set_text(day_name, -1)
                # Weekend headers slightly different color
                if col >= 5:
                    cr.set_source_rgb(*env.text_color_secondary)
                else:
                    cr.set_source_rgb(*env.text_color_primary)


def drawDaysGrid(cr, env, first_of_month):
    """Draw the days grid with date numbers and week numbers."""
    # Find first day position (Monday = 0)
    first_weekday = first_of_month.weekday()

    # Get number of days in month
    days_in_month = calendar.monthrange(first_of_month.year, first_of_month.month)[1]

    ONE_DAY = datetime.timedelta(days=1)
    date = first_of_month

    for day in range(1, days_in_month + 1):
        # Calculate cell position
        day_offset = first_weekday + day - 1
        col = day_offset % 7
        row = day_offset // 7

        x = env.offset_x + (col * env.cell_width)
        y = env.grid_offset_y + (row * env.cell_height)

        # Check if Monday (start of week) - show week number
        is_monday = date.weekday() == 0

        drawDayCell(cr, env, x, y, date, is_monday)
        date += ONE_DAY


def drawDayCell(cr, env, x, y, date, show_week_number):
    """Draw a single day cell with date number (top-right) and week number (top-left)."""
    with drawing.restoring_transform(cr):
        cr.translate(x, y)

        # Fill background for weekends
        if date.isoweekday() >= 6:
            cr.rectangle(0, 0, env.cell_width, env.cell_height)
            cr.set_source_rgba(*env.weekend_bg_color, 1.0)
            cr.fill()

        # Stroke cell border
        cr.set_source_rgba(*env.border_color, 1.0)
        cr.set_line_width(env.line_width)
        cr.rectangle(0, 0, env.cell_width, env.cell_height)
        cr.stroke()

        font_size = env.day_text_size
        padding = font_size * 0.5

        font = Pango.FontDescription("%s %s" % (env.font, font_size))
        font_bold = Pango.FontDescription("%s bold %s" % (env.font, font_size))

        # Draw day number in top-right corner
        day_str = str(date.day)

        day_layout = drawing.create_layout_with_kerning(cr)
        day_layout.set_font_description(font_bold)
        day_layout.set_text(day_str, -1)
        day_width, _ = day_layout.get_pixel_size()

        day_x = env.cell_width - padding - day_width

        with drawing.layout(cr) as layout:
            layout.set_font_description(font_bold)
            layout.set_text(day_str, -1)
            cr.move_to(day_x, padding)
            cr.set_source_rgb(*env.text_color_primary)

        # Draw week number in top-left corner (Monday only)
        if show_week_number:
            iso_week = date.isocalendar()[1]
            week_str = str(iso_week)

            with drawing.layout(cr) as layout:
                layout.set_font_description(font)
                layout.set_text(week_str, -1)
                cr.move_to(padding, padding)
                cr.set_source_rgb(*env.text_color_secondary)

File: src/layout/month.rs
use chrono::{Datelike, NaiveDate};

use crate::calendar;
use crate::canvas::{Canvas, Font};
use crate::config::{Config, Orientation, PageSetup, MM};
use crate::style::{defaults, Color};

pub struct MonthLayout;

impl super::Layout for MonthLayout {
    fn orientation(&self) -> Orientation {
        Orientation::Portrait
    }

    fn draw(&self, canvas: &Canvas, config: &Config, page: &PageSetup) {
        let avail_w = page.available_width();
        let avail_h = page.available_height();

        // Split page: 1/3 header, 2/3 grid
        let header_height = avail_h / 3.0;
        let grid_height = avail_h * 2.0 / 3.0;

        // Grid: 7 columns, 6 rows max, plus weekday header row
        let cols = 7;
        let header_row_h = grid_height * 0.08;
        let grid_content_h = grid_height - header_row_h;
        let rows = 6;
        let cell_w = avail_w / cols as f64;
        let cell_h = grid_content_h / rows as f64;

        let offset_x = page.margin;
        let header_y = page.margin;
        let weekday_y = page.margin + header_height;
        let grid_y = weekday_y + header_row_h;

        let line_width = defaults::BORDER_WIDTH_MM * MM;
        let day_font_size = 8.0;

        let date = NaiveDate::from_ymd_opt(config.year, config.month, 1).unwrap();

        draw_header(canvas, config, offset_x, header_y, avail_w, header_height);
        draw_weekday_headers(canvas, config, offset_x, weekday_y, cell_w, header_row_h);
        draw_days_grid(
            canvas, config, date, offset_x, grid_y, cell_w, cell_h, line_width, day_font_size,
        );
    }
}

fn draw_header(
    canvas: &Canvas,
    config: &Config,
    offset_x: f64,
    header_y: f64,
    avail_w: f64,
    header_height: f64,
) {
    canvas.with_save(|_| {
        let year_str = config.year.to_string();
        let date = NaiveDate::from_ymd_opt(config.year, config.month, 1).unwrap();
        let month_str = date.format("%b").to_string().to_uppercase();

        let month_font_size = (header_height * 0.5) as f64;
        let year_font_size = (month_font_size / 5.0) as f64;

        let year_font = Font::new(&config.font, year_font_size).bold();
        let month_font = Font::new(&config.font, month_font_size).bold();

        let year_m = canvas.measure_text(&year_str, &year_font);
        let month_m = canvas.measure_text(&month_str, &month_font);

        let center_x = offset_x + avail_w / 2.0;

        // Center month in header
        let month_y = header_y + (header_height - month_m.height) / 2.0;
        let month_x = center_x - month_m.width / 2.0;

        // Year above month
        let gap = year_m.height * 0.3;
        let year_y = (month_y - gap - year_m.height).max(header_y);
        let year_x = center_x - year_m.width / 2.0;

        canvas.draw_text(&year_str, year_x, year_y, &year_font, defaults::TEXT_PRIMARY);
        canvas.draw_text(
            &month_str,
            month_x,
            month_y,
            &month_font,
            defaults::TEXT_PRIMARY,
        );
    });
}

fn draw_weekday_headers(
    canvas: &Canvas,
    config: &Config,
    offset_x: f64,
    weekday_y: f64,
    cell_w: f64,
    header_row_h: f64,
) {
    let weekdays = ["MON", "TUE", "WED", "THU", "FRI", "SAT", "SUN"];
    let font_size = header_row_h * 0.5;
    let font = Font::new(&config.font, font_size).bold();

    for (col, name) in weekdays.iter().enumerate() {
        let x = offset_x + col as f64 * cell_w;

        canvas.with_save(|c| {
            c.translate(x, weekday_y);

            let m = canvas.measure_text(name, &font);
            let text_x = (cell_w - m.width) / 2.0;
            let text_y = (header_row_h - m.height) / 2.0;

            let color = if col >= 5 {
                defaults::TEXT_SECONDARY
            } else {
                defaults::TEXT_PRIMARY
            };
            c.draw_text(name, text_x, text_y, &font, color);
        });
    }
}

fn draw_days_grid(
    canvas: &Canvas,
    config: &Config,
    first_of_month: NaiveDate,
    offset_x: f64,
    grid_y: f64,
    cell_w: f64,
    cell_h: f64,
    line_width: f64,
    day_font_size: f64,
) {
    let first_weekday = first_of_month.weekday().num_days_from_monday() as usize;
    let days_in_month = calendar::last_day_of_month(config.year, config.month);

    let font = Font::new(&config.font, day_font_size);
    let font_bold = Font::new(&config.font, day_font_size).bold();
    let padding = day_font_size * 0.5;

    for day in 1..=days_in_month {
        let day_offset = first_weekday + day as usize - 1;
        let col = day_offset % 7;
        let row = day_offset / 7;

        let x = offset_x + col as f64 * cell_w;
        let y = grid_y + row as f64 * cell_h;

        let date = NaiveDate::from_ymd_opt(config.year, config.month, day).unwrap();
        let is_monday = date.weekday() == chrono::Weekday::Mon;

        canvas.with_save(|c| {
            c.translate(x, y);

            // Weekend background
            if calendar::is_weekend(date) {
                c.fill_rect(0.0, 0.0, cell_w, cell_h, defaults::WEEKEND_BG);
            }

            // Cell border
            c.stroke_rect(0.0, 0.0, cell_w, cell_h, defaults::BORDER, line_width);

            // Day number, top-right
            let day_str = day.to_string();
            let m = canvas.measure_text(&day_str, &font_bold);
            let day_x = cell_w - padding - m.width;
            c.draw_text(&day_str, day_x, padding, &font_bold, defaults::TEXT_PRIMARY);

            // Week number on Mondays, top-left
            if is_monday {
                let iso_week = date.iso_week().week().to_string();
                c.draw_text(&iso_week, padding, padding, &font, defaults::TEXT_SECONDARY);
            }
        });
    }
}
